introduce a new package censusdb

p4u opened the pull request:
The database operations previously implemented in the API census handler, are now moved to a standalone package.

The reasons are:
1) cleaner code since we do not mix API and database 
2) the censusdb can now be used by other components,
   such as the vochain/censusdownloader in order to
   import new census automatically when an election
   is created.

censusdb and api/census will deprecate censusManager

Signed-off-by: p4u <pau@dabax.net>

p4u committed:
introduce a new package censusdb

The database operations previously implemented in the
API census handler, are now moved to a standalone package.

The reasons are:
1) cleaner code since we do not mix API and database
2) the censusdb can now be used by other components,
   such as the vochain/censusdownloader in order to
   import new census automatically when an election
   is created.

censusdb and api/census will deprecate censusManager

Signed-off-by: p4u <pau@dabax.net>

p4u commented on api/api.go:
```diff
@@ -34,9 +34,8 @@ type API struct {
-
-	db db.Database
+	censusdb     *censusdb.CensusDB
+	db           db.Database // used for internal db operations
```
This database is currently used by the wallet handler. But ideally the wallet handler would have its own db so this "general" database can be removed from the API package.

mvdan commented on api/censusdb/censusdb.go:
```diff
@@ -0,0 +1,199 @@
+type CensusDB struct {
+	db           db.Database
+	censusMap    sync.Map
+	censusDBlock sync.RWMutex
```
what is this mutex for? note that pebble is safe for concurrent use, as long as each goroutine uses a separate batch (transaction), which is the case.

p4u replied:
Ummm, you are right, this lock might probably be removed. I'll check it.

mvdan commented on api/censusdb/censusdb.go:
```diff
@@ -0,0 +1,199 @@
+		ref.tree.Publish()
+	}
+	// add the tree in memory so we can quickly access to it afterwards
+	c.censusMap.Store(string(censusID), *ref)
```
I'm confused by censusMap. If we're OK with in-memory storage, we don't need pebble. If we want to persist the data across restarts, then I think censusMap is incomplete, because it doesn't load data from an existing database when dvotenode is started.

mvdan replied:
Also, assuming we want persistent storage, I think we could use pebble alone. As a key-value database which supports concurrent operations, one "get key" operation will be pretty fast. so I don't think a sync.Map is necessary.

p4u replied:
Using pebble alone would work. But why not use a in-memory data store? It adds a bit more of complexity to the code, but might have a positive impact on the performance. These queries are handled by the API, for a large census with thousands of voters, there might be many Read access. So having a copy in memory for those references that are actually used, looks like a good approach.

p4u replied:
Do you think Pebble cache will offer the same performance as having this sync.Map @mvdan ?

mvdan replied:
I'm willing to bet that brief reads on pebble are not going to be noticeably slow. I can see on their benchmarks that write operations execute at tens of thousands per second, and read operations at hundreds of thousands per second, for example.

Ultimately there aren't benchmarks or performance measurements here, so we're both just guessing. Which is why I'd personally go with less code until we can prove that there's a bottleneck that needs fixing. But I would be very surprised if the read throughput of pebble is not enough for our load.

mvdan commented on api/censusdb/censusdb.go:
```diff
@@ -0,0 +1,199 @@
+		if err != nil {
+			log.Warnf("error deleting census tree %x: %s", censusID, err)
+		}
+	}()
```
If we keep pebble alone, without mutexes, then I would delete this goroutine. you would not be preventing other operations by keeping the transaction open for a little longer, to delete the entire census tree. you'd also save work - you'd use one goroutine and one transaction, whereas now you use two goroutines and two transactions.

p4u replied:
The Delete operation might take many seconds (depending on the size of the census) since the only way to achieve it is by iterating over all tree leafs. This goroutine ensures the HTTP request is replied before any timeout is triggered. And once the reference is removed (that's executed before launching the routine) at practical effects the census is not available anymore for the API.

mvdan commented on api/censusdb/censusdb.go:
```diff
@@ -0,0 +1,199 @@
+		}
+		return nil, err
+	}
+	dec := gob.NewDecoder(bytes.NewBuffer(b))
```
nit: bytes.NewReader is enough

p4u committed:
move compressor to a separate package so it can be reused

Signed-off-by: p4u <pau@dabax.net>

mvdan commented on api/censusdb/censusdb.go:
```diff
@@ -0,0 +1,199 @@
+	IsPublic   bool
+}
+
+// Tree returns the censustree.Tree object of the census reference.
+func (cr *CensusRef) Tree() *censustree.Tree {
+	return cr.tree
+}
+
+// SetTree sets the censustree.Tree object to the census reference.
+func (cr *CensusRef) SetTree(tree *censustree.Tree) {
+	cr.tree = tree
+}
```
just make the tree field public :) to avoid encoding it in the gob, you could do something like:

```
type CensusRef struct {
    Tree ...
    Encoded EncodedCensusRef
}

type EncodedCensusRef struct {
    AuthToken ...
    ...
}
```

Then you only encode and decode the `Encoded` field. No real difference in terms of memory, but you do avoid the get/set methods.

p4u replied:
Nice trick. I was looking for a struct tag able to exclude a value from gob but it does not exist. However I like more the get/set approach since we are only avoiding 1 field from serialization (fore more, I'd see your approach better).

p4u committed:
censusdb: remove lock

Since Pebble db is already safe and we use a sync.Map
for in-memory data, we don't need a Mutex lock.

Signed-off-by: p4u <pau@dabax.net>

p4u committed:
censusdb: remove memory map and improve error handling

Signed-off-by: p4u <pau@dabax.net>

p4u committed:
use compressor from data package

Signed-off-by: p4u <pau@dabax.net>
